src/widgets/concept-widget.tsx

```diff
@@ -0,0 +1,148 @@
+import {
+  VDomModel, VDomRenderer
+} from '@jupyterlab/apputils';
+import {INotebookTracker} from '@jupyterlab/notebook';
+
+import * as React from 'react';
+
+import {AllOfUsConfig} from '../config';
+import {Concept} from '../generated';
+import {ConceptsService} from '../services/concepts.service';
+
+import {Observable} from 'rxjs/Rx';
+
+export enum LoadingStatus {
+  UNSET,
+  LOADING,
+  LOADED
+}
+
+export class ConceptsWidgetModel extends VDomModel {
+
+  public initialized: boolean;
+  public loadingStatus = LoadingStatus.UNSET;
+  public query = '';
+  public concepts: Array<Concept>;
+  public selectedConcept: Concept;
+}
+
+export class ConceptsWidgetRenderer extends VDomRenderer<ConceptsWidgetModel> {
+
+
+  constructor(configObservable: Observable<AllOfUsConfig>,
+              private conceptsService: ConceptsService,
+              private notebooks: INotebookTracker) {
+    super();
+
+    this.id = 'allofus-concepts';
+    this.title.label = 'Concepts';
+    this.title.closable = true;
+    configObservable.subscribe((config) => {
+      if (!this.model.initialized) {
+        this.model.initialized = true;
+        this.update();
+      }
+    });
+  }
+
+  protected render(): React.ReactElement<any> {
+    const handleQueryChange = (event) => {
+      this.model.query = event.target.value;
+      this.update();
+    };
+
+    const searchConcepts = () => {
+      this.model.loadingStatus = LoadingStatus.LOADING;
+      this.update();
+      this.conceptsService.searchConcepts({query: this.model.query})
+          .then((concepts) => {
+            this.model.concepts = concepts;
+            this.model.loadingStatus = LoadingStatus.LOADED;
+            this.update();
+          });
+    };
+    return (
+        <div>
```

Comment:
I'm less OK with this approach. One of the strengths of React's virtual DOM is to allow React to manage the re-rendering. If some element of state changes inside of a component, then React will know to call render for the specific elements that are touched by that element of state. The approach here subsumes that primary React function and takes control over the whole dom.

Reply:
I'm revising my critique here. After reviewing this with Mohs, I think I'm wrong - we still are letting React do the rendering (which it will optimize via the virtual DOM) even if we tell it to re-render the whole table.